docs/utilities/idempotency.md

```diff
@@ -77,31 +77,68 @@ If you're not [changing the default configuration for the DynamoDB persistence l
     You can reuse the same DynamoDB table to store idempotency state. We add `module_name` and [qualified name for classes and functions](https://peps.python.org/pep-3155/){target="_blank"} in addition to the idempotency key as a hash key.
 
-```yaml hl_lines="5-13 21-23" title="AWS Serverless Application Model (SAM) example"
-Resources:
-  IdempotencyTable:
-	Type: AWS::DynamoDB::Table
-	Properties:
-	  AttributeDefinitions:
-		-   AttributeName: id
-			AttributeType: S
-	  KeySchema:
-		-   AttributeName: id
-			KeyType: HASH
-	  TimeToLiveSpecification:
-		AttributeName: expiration
-		Enabled: true
-	  BillingMode: PAY_PER_REQUEST
-
-  HelloWorldFunction:
-  Type: AWS::Serverless::Function
-  Properties:
-	Runtime: python3.9
-	...
-	Policies:
-	  - DynamoDBCrudPolicy:
-		  TableName: !Ref IdempotencyTable
-```
+=== "sam.yaml"
+
+    ```yaml hl_lines="5-13 21-23" title="AWS Serverless Application Model (SAM) example"
+    Resources:
+    IdempotencyTable:
+        Type: AWS::DynamoDB::Table
+        Properties:
+        AttributeDefinitions:
+            -   AttributeName: id
+                AttributeType: S
+        KeySchema:
+            -   AttributeName: id
+                KeyType: HASH
+        TimeToLiveSpecification:
+            AttributeName: expiration
+            Enabled: true
+        BillingMode: PAY_PER_REQUEST
+
+    HelloWorldFunction:
+    Type: AWS::Serverless::Function
+    Properties:
+        Runtime: python3.9
+        ...
+        Policies:
+        - DynamoDBCrudPolicy:
+            TableName: !Ref IdempotencyTable
+    ```
+=== "cdk.py"
+
+    ```python hl_lines="14 17 26-29"  title="AWS Cloud Development Kit (CDK) Construct example"
+    from aws_cdk import RemovalPolicy
+    from aws_cdk import aws_dynamodb as dynamodb
+    from aws_cdk import aws_iam as iam
+    from constructs import Construct
+
+
+    class IdempotencyConstruct(Construct):
+
+        def __init__(self, scope: Construct, id_: str, lambda_role: iam.Role) -> None:
+            super().__init__(scope, id_)
+            self.idempotency_table = dynamodb.Table(
+                self,
+                'IdempotencyTable',
+                partition_key=dynamodb.Attribute(name='id', type=dynamodb.AttributeType.STRING),
+                billing_mode=dynamodb.BillingMode.PAY_PER_REQUEST,
+                removal_policy=RemovalPolicy.DESTROY,
+                time_to_live_attribute='expiration',
+                point_in_time_recovery=True,
+            )
+            lambda_role.attach_inline_policy(
+            iam.Policy(
+                self,
+                'idempotency-policy',
+                statements=[
+                    iam.PolicyStatement(
+                        actions=['dynamodb:PutItem', 'dynamodb:GetItem', 'dynamodb:UpdateItem', 'dynamodb:DeleteItem'],
+                        resources=[self.idempotency_table.table_arn],
+                        effect=iam.Effect.ALLOW,
+                    )
+                ]
+            ))
```

Comment:
Maybe we could use the `grant` functionality of CDK instead?
```suggestion
            self.idempotency_table.grant_read_write_data(lambda_role)
```

Reply:
i dotn usually recommend using the grant_read_write_data as it provides more permissions than you really need (I checked). i wanted it to be least privilege.

Reply:
I understand. Two asks then:

**1/** Use `dynamodb.grant` with the permissions you need, as it's more CDK friendly:

```suggestion
            self.idempotency_table.grant(lambda_role, ['dynamodb:PutItem', 'dynamodb:GetItem', 'dynamodb:UpdateItem', 'dynamodb:DeleteItem'])
```

**2/** We will need to update the SAM example for consistency, as it currently uses the overly-permissive `DynamoDBCrudPolicy`.

Let me know if you need a hand for this!